drivers/firmware/scmi/core.c

```diff
@@ -215,6 +215,47 @@ int scmi_send_message(struct scmi_protocol *proto, struct scmi_message *msg,
 	}
 }
 
+static int scmi_core_protocol_negotiate(struct scmi_protocol *proto)
+{
+	uint32_t agent_version, protocol_version;
+	int ret;
+
+	if (!proto) {
+		return -EINVAL;
+	}
+
+	protocol_version = scmi_protocol_get_supported_version(proto);
+
+	if (protocol_version) {
+		ret = scmi_protocol_version_get(proto, &agent_version);
+		if (ret < 0) {
+			LOG_ERR("Protocol 0x%X: Failed to get agent version: %d",
+				proto->id, ret);
+			return ret;
+		}
+
+		if (agent_version > protocol_version) {
+			LOG_INF("Protocol 0x%X: Platform version (0x%08x) higher than "
+				"supported (0x%08x), negotiating...",
+				proto->id, agent_version, protocol_version);
+
+			ret = scmi_protocol_version_negotiate(proto, protocol_version);
+			if (ret == 0) {
+				LOG_INF("Protocol 0x%X: Successfully negotiated to version 0x%08x",
+					proto->id, protocol_version);
+			} else {
+				LOG_ERR("Protocol 0x%X: Negotiation failed (%d)", proto->id, ret);
+				return ret;
+			}
+		} else {
```

Comment:
The case when our version is higher than the platform's exposed version (firmware is too old) seems to falsely fall into the compatible category.